PP-482: Soft Walltime

bhroam opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-482](https://pbspro.atlassian.net/browse/PP-482)**
* **[PP-938](https://pbspro.atlassian.net/browse/PP-938)**
* **[PP-853](https://pbspro.atlassian.net/browse/PP-853)**

#### Problem description
* PP-482: RFE non-destructive walltime
* PP-938: Action functions do not handle interdependent value validity set at the same time
* PP-853: If a STF job is qaltered where max_walltime < min_walltime, both are unset

#### Cause / Analysis
* PP-482 adds a resource similar to walltime (soft_walltime) which will only be used by the scheduler.  Since it is only used by the scheduler, the job will not be killed if it is exceeded.
* PP-938: If two interdependent resources are set at the same time, the action functions don't function properly.  An action function is passed in the new attribute and the object it is being set on.  Since setting attributes is an atomic operation, the attributes aren't set until after all the action functions pass.  This means if any attribute is queried from the object, it is pointing to the old value.  It leads to situations like soft_walltime > walltime
* PP-853: STF jobs had special case code in modify_job_attr().  It would set both min and max walltime.  It would make checks (like min_walltime < max_walltime).  If any check failed, the resources would be unset.  If either resource was set previously, it would be unset afterwards.

#### Solution description
* PP-482: This introduces a new resource called soft_walltime.  When this resource is set, the scheduler sets the job's duration to this value instead of the walltime.  If a running job's resources_used.walltime exceeds its current soft_walltime, the soft_walltime is extended by 100% of its original soft_walltime request.
* PP-938: This affects modify_job_attr(qalter) and mgr_attr_set2 (qmgr).  The new method will save a copy of the object's original attributes.  It then applies all the values being set.  The action functions are called.  If any action function fails, the saved copy is put back in place.
* PP-853: I moved the special case code out of modify_job_attr() into the action functions for min_walltime and max_walltime.  This is the right place for these types of checks to be made.  Also, I needed to add more checks because STF jobs can not request soft_walltime.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

minghui-liu commented on src/scheduler/misc.c:
```diff
@@ -814,22 +814,32 @@ calc_time_left_STF(resource_resv *resresv, sch_resource_t* min_time_left)
+	long duration;
 
-	if (resresv->duration == UNSPECIFIED)
+	if(use_hard_duration && resresv->hard_duration == UNSPECIFIED)
```
Please add a space after `if`.

minghui-liu commented on src/scheduler/misc.c:
```diff
@@ -814,22 +814,32 @@ calc_time_left_STF(resource_resv *resresv, sch_resource_t* min_time_left)
+	else if (!use_hard_duration && resresv->duration == UNSPECIFIED)
+		return -1;
+
+	if(use_hard_duration)
```
Please add a space after `if`.

minghui-liu commented on src/scheduler/misc.c:
```diff
@@ -814,22 +814,32 @@ calc_time_left_STF(resource_resv *resresv, sch_resource_t* min_time_left)
 	used_amount = calc_used_walltime(resresv);
 
-	return IF_NEG_THEN_ZERO((resresv->duration - used_amount));
+	return IF_NEG_THEN_ZERO((duration - used_amount));
```
`IF_NEG_THEN_ZERO` will wrap argument with `()` so the inner pair of `()` is unnecessary.

minghui-liu commented on src/scheduler/misc.h:
```diff
@@ -246,23 +246,22 @@ res_to_str_re(void *p, enum resource_fields fld, char **buf,
+void move_schd_error(schd_error *err, schd_error *oerr);
+
+/* do a deep copy of err, but don't dup the error itself. */
+void  copy_schd_error(schd_error *err, schd_error *oerr);
```
Please remove one space after `void`.

minghui-liu commented on src/scheduler/node_info.c:
```diff
@@ -3704,14 +3704,18 @@ check_resources_for_node(resource_req *resreq, node_info *ninfo,
 	calendar = ninfo->server->calendar;
 	cur_time = ninfo->server->server_time;
-	end_time = cur_time + calc_time_left(resresv);
+	if(resresv->duration != resresv->hard_duration &&
```
Please add a space after `if`.

minghui-liu commented on src/scheduler/node_info.c:
```diff
@@ -5437,7 +5441,12 @@ sim_exclhost(event_list *calendar, resource_resv *resresv, node_info *ninfo)
 		return 1;
 
-	end = resresv->server->server_time + calc_time_left(resresv);
+	if(resresv->duration != resresv->hard_duration &&
```
Please add a space after `if`.

minghui-liu commented on src/scheduler/resv_info.c:
```diff
@@ -230,7 +230,17 @@ query_reservations(server_info *sinfo, struct batch_status *resvs)
+		 * to confirm reservations.  In that case, build the calendar
+		 * using the hard durations of jobs.
+		 */
+		if(will_confirm(resresv, sinfo->server_time))
```
Please add a space after `if`.

minghui-liu commented on src/scheduler/simulate.c:
```diff
@@ -637,6 +637,39 @@ exists_run_event(event_list *calendar, time_t end)
+	timed_event *te;
+	timed_event *te_list;
+
+	if(calendar == NULL)
```
Please add a space after `if`.

minghui-liu commented on src/scheduler/simulate.c:
```diff
@@ -637,6 +637,39 @@ exists_run_event(event_list *calendar, time_t end)
+	for (te = te_list; te != NULL && te->event_time <= end;
+		te = find_next_timed_event(te, 0, TIMED_RUN_EVENT)) {
+		if (te->event_type == TIMED_RUN_EVENT) {
+			resource_resv *resresv = (resource_resv *) te->event_ptr;
```
Please remove space after type cast.

minghui-liu commented on src/scheduler/simulate.c:
```diff
@@ -637,6 +637,39 @@ exists_run_event(event_list *calendar, time_t end)
+		te = find_next_timed_event(te, 0, TIMED_RUN_EVENT)) {
+		if (te->event_type == TIMED_RUN_EVENT) {
+			resource_resv *resresv = (resource_resv *) te->event_ptr;
+			if(resresv->is_resv)
```
Please add a space after `if`.

minghui-liu commented on src/server/req_manager.c:
```diff
@@ -921,11 +932,83 @@ mgr_set_attr2(attribute *pattr, attribute_def *pdef, int limit, svrattrl *plist,
 		}
 	}
 
+	for(index = 0; index < limit; index++) {
```
Please add a space after `for`.

minghui-liu commented on src/server/req_modify.c:
```diff
@@ -498,10 +469,37 @@ modify_job_attr(job *pjob, svrattrl *plist, int perm, int *bad)
+	 */
 
+	newattr = calloc(JOB_ATR_LAST, sizeof(attribute));
+	if(newattr == NULL)
```
Please add a space after `if`.

minghui-liu commented on src/server/req_modify.c:
```diff
@@ -498,10 +469,37 @@ modify_job_attr(job *pjob, svrattrl *plist, int perm, int *bad)
+		return PBSE_SYSTEM;
 	rc = attr_atomic_set(plist, pattr, newattr, job_attr_def, JOB_ATR_LAST,
 		allow_unkn, perm, bad);
+	if(rc) {
```
Please add a space after `if`.

minghui-liu commented on src/server/req_modify.c:
```diff
@@ -498,10 +469,37 @@ modify_job_attr(job *pjob, svrattrl *plist, int perm, int *bad)
+	}
+
+	newattr2 = calloc(JOB_ATR_LAST, sizeof(attribute));
+	if(newattr2 == NULL) {
```
Please add a space after `if`.

minghui-liu commented on src/server/req_modify.c:
```diff
@@ -498,10 +469,37 @@ modify_job_attr(job *pjob, svrattrl *plist, int perm, int *bad)
+	attr_atomic_copy(newattr2, newattr, job_attr_def, JOB_ATR_LAST);
+
+	attr_save = calloc(JOB_ATR_LAST, sizeof(attribute));
+	if(attr_save == NULL) {
```
Please add a space after `if`.

minghui-liu commented on src/server/req_modify.c:
```diff
@@ -650,16 +634,58 @@ modify_job_attr(job *pjob, svrattrl *plist, int perm, int *bad)
+		}
+	}
+
+	for (i = 0; i<JOB_ATR_LAST; i++) {
```
Please add spaces around `<`.

minghui-liu commented on src/server/req_manager.c:
```diff
@@ -810,64 +813,72 @@ mgr_set_attr2(attribute *pattr, attribute_def *pdef, int limit, svrattrl *plist,
+	attr_atomic_copy(pre_copy, new, pdef, limit);
+
+	attr_save = calloc((unsigned int)limit, sizeof(attribute));
+	if(attr_save == (attribute *) 0) {
```
Please add a space after `if`.

minghui-liu commented on src/server/req_manager.c:
```diff
@@ -810,64 +813,72 @@ mgr_set_attr2(attribute *pattr, attribute_def *pdef, int limit, svrattrl *plist,
 	}
 
+	pre_copy = (attribute *)calloc((unsigned int)limit, sizeof(attribute));
+	if (new == (attribute *)0) {
```
You meant to check `pre_copy` instead of `new` right?

minghui-liu commented on src/server/req_manager.c:
```diff
@@ -810,64 +813,72 @@ mgr_set_attr2(attribute *pattr, attribute_def *pdef, int limit, svrattrl *plist,
 					if ((pleaf->slf_limit.at_flags & (ATR_VFLAG_SET|ATR_VFLAG_MODIFY)) == ATR_VFLAG_MODIFY) {
 						*bad = index + 1;
 						attr_atomic_kill(new, pdef, limit);
+						attr_atomic_kill(pre_copy, pdef, limit);
```
What about `attr_save`. I think it also needs to be freed.

minghui-liu commented on src/server/resc_attr.c:
```diff
@@ -907,3 +908,151 @@ int apply_select_inchunk_rules(resource *presc, attribute *pattr, void *pobj, in
+		}
+	}
+	return PBSE_NONE;
+}
```
Please add a space after all the`if`s in this function.

minghui-liu commented on src/server/resc_attr.c:
```diff
@@ -907,3 +908,151 @@ int apply_select_inchunk_rules(resource *presc, attribute *pattr, void *pobj, in
+		}
+	}
+	return PBSE_NONE;
+}
```
Please add a space after all the`if`s in this function.

arungrover commented on test/tests/functional/pbs_soft_walltime.py:
```diff
@@ -0,0 +1,634 @@
+                elif 'Resource_List.walltime' in Jstat[0]:
+                    wt2 = int(Jstat[0]['Resource_List.walltime'])
+
+            self.assertTrue(est + wt, est2)
```
Like you said, this looks like a problem

arungrover commented on test/tests/functional/pbs_soft_walltime.py:
```diff
@@ -0,0 +1,634 @@
+
+        J = Job(TEST_USER, {ATTR_h: 'u'})
+        jid = self.server.submit(J)
+        self.server.alterjob(jid, {'Resource_List.soft_walltime': 240})
```
why not submit the job with soft_walltime instead of altering it?

bhroam replied:
Job submissions happen with user perms.  Even if the user has manager perms, they still can't submit a job requesting soft_walltime.  I'm not exactly sure for the reason of this.  Maybe we can ask Bob the next time we see him.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -732,31 +741,60 @@ query_jobs(status *policy, int pbs_sd, queue_info *qinfo, resource_resv **pjobs,
+			/* Duration has been exceeded - either extend soft_walltime or expect the job to be killed */
+			else {
+				if (soft_walltime_req != NULL) {
+					duration = extend_soft_walltime(resresv, server_time);
```
I removed the UNSPECIFIED return value and made it JOB_INFINITY like it should have been

minghui-liu replied:
Since the else branch has multiple lines please wrap the other branches with brackets as well.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -5068,3 +5105,54 @@ resource_req *create_resreq_rel_list(status *policy, nspec **res_released)
+	num_ext_over = (server_time - resresv->job->stime) / job_duration;
+
+	extension = num_ext_over * job_duration;
+	extended_duration = job_duration + extension;
```
It's doing the right thing.  We take the current time and subtract the start time of the job.  This is the current run time.  We divide that by the soft_walltime to get num_ext_over.  For example if the soft_walltime is 60 and the run time is 180, num_ext_over is 3.  Now this is just the number of times we've exceeded the soft_walltime estimate.  We need to add one more soft_walltime to correctly extend it.

arungrover replied:
ah ok, thanks for clarification. I misread it.

bhroam commented on src/scheduler/misc.c:
```diff
@@ -814,22 +814,32 @@ calc_time_left_STF(resource_resv *resresv, sch_resource_t* min_time_left)
+	if (use_hard_duration && resresv->hard_duration == UNSPECIFIED)
 		return -1;
 
+	else if (!use_hard_duration && resresv->duration == UNSPECIFIED)
```
If a job is not submitted with walltime or soft_walltime both duration and hard_duration are set to JOB_INFINITY which is 5 years.

bhroam commented on src/server/req_manager.c:
```diff
@@ -810,64 +813,73 @@ mgr_set_attr2(attribute *pattr, attribute_def *pdef, int limit, svrattrl *plist,
 		return (rc);
 	}
 
+	pre_copy = (attribute *)calloc((unsigned int)limit, sizeof(attribute));
```
Once you copy the values from one attribute array to another, all resource and list types are no longer there.  You've done a shallow copy from one to the other.  If we were apply new to the object, we couldn't call the action functions with new since all resources would no longer be there.  This is what pre_copy is for.
We do the following:
1) copy pre_copy to pattr (the object's real attributes)
2) call the action functions with new
3) If any action function fails, set pattr back to attr_save
4) If the action functions are successful, we need to copy new to pattr since the action functions might have changed new.

There's a comment there trying to explain this.  If you have any suggestions on how to make it better, please help.

bhroam commented on src/server/req_manager.c:
```diff
@@ -921,11 +933,83 @@ mgr_set_attr2(attribute *pattr, attribute_def *pdef, int limit, svrattrl *plist,
+
+	}
+
+	/* The action functions might have modified new2.  Need to set pattr again */
```
Typo, it should have been new.

bhroam commented on src/server/req_manager.c:
```diff
@@ -921,11 +933,83 @@ mgr_set_attr2(attribute *pattr, attribute_def *pdef, int limit, svrattrl *plist,
+			if (pold->at_type == ATR_TYPE_LIST) {
+				list_move(&pnew->at_val.at_list, &pold->at_val.at_list);
+			} else if (pold->at_type == ATR_TYPE_RESC) {
+				set_resc(pold, pnew, INCR);
```
What you say makes sense to me.  I don't understand why it is being called with INCR.  I am just following what was originally done.  I'd worry it'd break if we change INCR to SET.

arungrover commented on src/server/resc_attr.c:
```diff
@@ -923,35 +935,47 @@ action_soft_walltime(resource *presc, attribute *pattr, void *pobject, int type,
-		def = find_resc_def(svr_resc_def, WALLTIME, svr_resc_size);
-		if (!def)
+		if (soft_walltime_def == NULL)
+			soft_walltime_def = find_resc_def(svr_resc_def, WALLTIME, svr_resc_size);
```
you are using wrong resource name here

arungrover commented on src/server/resc_attr.c:
```diff
@@ -908,13 +908,25 @@ int apply_select_inchunk_rules(resource *presc, attribute *pattr, void *pobj, in
 	if (pobject != NULL) {
-		resource_def *def;
+		static resource_def *soft_walltime_def = NULL;
+		static resource_def *walltime_def = NULL;
```
This variable is not needed

arungrover commented on src/server/resc_attr.c:
```diff
@@ -923,35 +935,47 @@ action_soft_walltime(resource *presc, attribute *pattr, void *pobject, int type,
 		/* soft_walltime and STF jobs are incompatible */
-		def = find_resc_def(svr_resc_def, MIN_WALLTIME, svr_resc_size);
-		if (!def)
+		if (min_walltime_def != NULL)
```
This should be - if (min_walltime_def == NULL)

arungrover commented on src/server/resc_attr.c:
```diff
@@ -960,27 +984,40 @@ action_walltime(resource *presc, attribute *pattr, void *pobject, int type, int
 		/* Make sure walltime > soft_walltime */
-		def = find_resc_def(svr_resc_def, SOFT_WALLTIME, svr_resc_size);
-		if (!def)
+		if (soft_walltime_def != NULL)
```
you want to check for soft_walltime_def to be NULL here

arungrover commented on src/server/resc_attr.c:
```diff
@@ -994,35 +1031,50 @@ action_min_walltime(resource *presc, attribute *pattr, void *pobject, int type,
 		/* max_walltime needs to be greater than min_walltime */
-		def = find_resc_def(svr_resc_def, MAX_WALLTIME, svr_resc_size);
-		if (!def)
+		if (max_walltime_def != NULL)
```
similar change needed here too

arungrover commented on src/server/resc_attr.c:
```diff
@@ -994,35 +1031,50 @@ action_min_walltime(resource *presc, attribute *pattr, void *pobject, int type,
-		if (!def)
+		if (soft_walltime_def == NULL)
+			soft_walltime_def = find_resc_def(svr_resc_def, SOFT_WALLTIME, svr_resc_size);
+		if (soft_walltime_def == NULL)
```
I'm not sure if you want to return from here... Although it is almost impossible to have a NULL return for these in-built resources but the action function is for min_walltime. I think we should continue forward to compare it with max_walltime

arungrover commented on src/server/resc_attr.c:
```diff
@@ -923,35 +935,47 @@ action_soft_walltime(resource *presc, attribute *pattr, void *pobject, int type,
-		if (!def)
+		if (soft_walltime_def == NULL)
+			soft_walltime_def = find_resc_def(svr_resc_def, WALLTIME, svr_resc_size);
+		if (soft_walltime_def == NULL)
```
This is an unlikely case but I think we should return something like PBSE_NORESC. Unfortunately that error message is not there, so the closest thing to it is PBSE_NOATTR.

bhroam replied:
I followed what was being done for STF jobs here.  I have an alternate solution.  How about if the resource def can't be found, we just skip that check (not return success).  I think this is the right thing to do.  The only way to get a NULL is if someone were to remove a resource.  Since these checks enforce a dependency between two resources, that dependency can't fail if one resource doesn't exist.

arungrover commented on src/server/resc_attr.c:
```diff
@@ -1031,24 +1083,25 @@ action_max_walltime(resource *presc, attribute *pattr, void *pobj, int type, int
 		/* max_walltime needs to be greater than min_walltime */
-		def = find_resc_def(svr_resc_def, MIN_WALLTIME, svr_resc_size);
-		if (!def)
+		min_walltime_def = find_resc_def(svr_resc_def, MIN_WALLTIME, svr_resc_size);
```
You want to check for min_walltime_def to be NULL and then do a find

arungrover commented on src/server/resc_attr.c:
```diff
@@ -1031,24 +1083,25 @@ action_max_walltime(resource *presc, attribute *pattr, void *pobj, int type, int
-		def = find_resc_def(svr_resc_def, MIN_WALLTIME, svr_resc_size);
-		if (!def)
+		min_walltime_def = find_resc_def(svr_resc_def, MIN_WALLTIME, svr_resc_size);
+		if (min_walltime_def != NULL)
```
You want reverse of this comparison, right?

arungrover commented on src/server/req_manager.c:
```diff
@@ -810,64 +813,73 @@ mgr_set_attr2(attribute *pattr, attribute_def *pdef, int limit, svrattrl *plist,
+	 * of attr_atomic_set() to '1'; otherwise, set it to '-1' meaning
+	 * not to allow unknown resource
+	 */
+	if ((rc = attr_atomic_set(plist, pattr, new, pdef, limit, (allow_unkresc?1:-1), privil, bad)) != 0) {
```
There seems to be a problem here, if we allow unknown resources then we should pass the last index value where we want to append the resource. '1' will probably make attr_atomic_set use wrong resource index (probably 'mem' resource) to decode/set the resource.

anamikau commented on test/tests/functional/pbs_soft_walltime.py:
```diff
@@ -0,0 +1,634 @@
+        """
+        Test that there is no change in the reservation behavior with
+        soft_walltime set on jobs with a hard walltime set.  The soft_walltime
+        should be ignored and only the hard walltime should be used.
```
It is not clear from the test that reservation gets deleted due to hard walltime. The behavior is exactly the same as test above where you were only setting soft walltime. May be you edit the test and set walltime to 59 and start reservation at now +60 and make sure it starts.

bhroam replied:
The two tests are testing distinct behavior.  One is testing a job with soft_walltime and no walltime.  The soft_walltime should not be used and the job should be considered taking forever.  The other is testing a job with soft_walltime and walltime, but the walltime is longer than the reservation start time.  This one proves that the soft_walltime is ignored and the walltime is used.  The test you are suggesting is just testing existing behavior.  It should exist in TestReservations, but doesn't.

anamikau replied:
Bhroam, I get the test cases. In first case if you do not specify the walltime will it be taking default, i.e., 5years?

anamikau commented on test/tests/functional/pbs_soft_walltime.py:
```diff
@@ -0,0 +1,634 @@
+    being extended properly when exceeded
+    """
+
+    def compare_estiamtes(self, running_job, jobs):
```
Why this function name contains 'estamtes' than 'estimates'? This looks wierd

bhroam replied:
It was a typo.  I fixed it.

anamikau commented on test/tests/functional/pbs_soft_walltime.py:
```diff
@@ -0,0 +1,634 @@
+        self.server.alterjob(jid3, {'Resource_List.soft_walltime': 150})
+
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid3)
```
please also check if job1 is running too

bhroam replied:
Looks like I missed that.  Thanks for catching it.

arungrover commented on src/server/resc_attr.c:
```diff
@@ -1033,24 +1032,23 @@ action_min_walltime(resource *presc, attribute *pattr, void *pobject, int type,
-		if (entry != NULL) {
-			if (entry->rs_value.at_flags & ATR_VFLAG_SET)
-				return PBSE_SOFTWT_STF;
+		if (soft_walltime_def == NULL) {
```
This needs to be soft_walltime_def != NULL

arungrover commented on src/server/resc_attr.c:
```diff
@@ -1033,24 +1032,23 @@ action_min_walltime(resource *presc, attribute *pattr, void *pobject, int type,
-			if (max_walltime_def->rs_comp(&(entry->rs_value), &(presc->rs_value)) < 0)
-				return PBSE_MIN_GT_MAXWT;
+			max_walltime_def = find_resc_def(svr_resc_def, MAX_WALLTIME, svr_resc_size);
+		if (max_walltime_def == NULL) {
```
this needs to be max_walltime_def != NULL

arungrover commented on src/server/resc_attr.c:
```diff
@@ -1085,27 +1083,27 @@ action_max_walltime(resource *presc, attribute *pattr, void *pobj, int type, int
-			return PBSE_MAX_NO_MINWT;
+		if (min_walltime_def == NULL)
+			min_walltime_def = find_resc_def(svr_resc_def, MIN_WALLTIME, svr_resc_size);
+		if (min_walltime_def == NULL) {
```
This needs to be min_walltime_def != NULL

bhroam commented on test/tests/functional/pbs_soft_walltime.py:
```diff
@@ -0,0 +1,634 @@
+    def test_soft_walltime_perms(self):
+        """
+        Test to see if soft_walltime can't be submitted with a job or
+        altered by a normal user
```
While I can do this, it is not necessary.  All jobs are submitted with user permissions.  If it can't be submitted as a normal user, it can't be submitted by one with higher permissions.  Do you still think it is necessary?

anamikau replied:
Yes design explicitly talks about operator permissions so it should also be checked explicitly checked.

bhroam replied:
@anamikau thank you for pointing this out.  It was an issue with my EDD.  I have updated the EDD to say that soft_walltime can not be requested at submit time.  With the update to the EDD, I don't think this test needs to be changed.  The fact that you can't request soft_walltime at submit time is because of existing behavior (all job submits use user permissions).

anamikau replied:
Thanks for updating the EDD for job submission. I still think we need to extend the alterjob case below. I will include it in QA scenarios.

bhroam commented on test/tests/functional/pbs_soft_walltime.py:
```diff
@@ -0,0 +1,634 @@
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+
+        self.server.expect(JOB, {'estimated.soft_walltime':
+                                 (MATCH_RE, '00:00:14|14')}, id=jid)
```
Will do.

bhroam commented on test/tests/functional/pbs_soft_walltime.py:
```diff
@@ -0,0 +1,634 @@
+        """
+
+        a = {'resources_available.ncpus': 4}
+        self.server.create_vnodes(name='vn', attrib=a, num=1, mom=self.mom)
```
Sure.

anamikau commented on test/tests/functional/pbs_soft_walltime.py:
```diff
@@ -0,0 +1,686 @@
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+
+        self.server.expect(JOB, {'estimated.soft_walltime':
+                                 (MATCH_RE, '00:00:14|14')}, id=jid)
```
Looking at it again I am thinking there might be a race condition on line 348. we are checking after 12s and job is supposed to be killed by 14s. 2s (though it is enough) but might possibly introduce some race conditions. I think we should change the walltime to 17-18s.

bhroam commented:
The walltime of 14 probably won't be an issue because PBS doesn't kill the job over its walltime limit right on time.  It waits until its next poll cycle.  I'm currently seeing it taking about 10s longer.  In any case, I updated the soft_walltime to 6 and the walltime to 16.  This gives 4 seconds between the second update and when it gets over its limit.  As for the other test cases, If you do not specify a walltime, it uses a virtual "infinity" as the job duration.  It just so happens the scheduler is using 5 years.  It isn't really a walltime.  If a job really took 5 years, it would not be killed.

bhroam commented:
I fixed some soft walltime PTL issues that @anamikau found.  I also added a test plan for STF.  I did this because I needed to write a test for PP-853.  I pulled over the production test plan and added to it.  In the process I updated the test plan. I basically fixed everything I'd comment on.   Feel free to review the whole test plan, or just test_t6.  That's the new one I wrote.  The rest of the test plan has already been reviewed when STF was originally checked in.  There is one test that is failing due to a PBS issue.  I'll file this.

arungrover commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+        with R117, R121, R124 having ncpus=3, all others having ncpus=2.
+        Duration of 10 min with 30 min difference between consecutive
+        reservation.	A STF job will shrink its walltime to less than or
+        equal to 3 hours.
```
it should be 4 hours here

arungrover commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+
+        qname = 'highp'
+
+        queues = self.server.status(QUEUE)
```
this stat queue is not needed

arungrover commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+        self.server.submit(r)
+
+        m = self.server.log_match(
+            "reservation deleted", starttime=(now - 1), max_attempts=10)
```
if starttime is now -1 then wouldn't it just match the previous denial of reservation. I think we nee to sleep for a second and then set now again

bhroam replied:
I removed all the starttime parameters to log_match() and added the reservation id.  The log message is actually '<rid>;reservation deleted'

anamikau commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
```
Should we remove the commercial license information since it is going in OSS.

bhroam replied:
This is the OSS license.  I copied it from another test.  PBS is under a dual-license (OSS and commercial).

anamikau commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+        msg = "Job;%s;Job will run for duration=[%s|%s]" % (j2id, wt, wt2)
+        m = self.scheduler.log_match(
+            msg, regexp=True, starttime=(now - 1), max_attempts=5, interval=2)
+        self.assertTrue(len(m) > 0)
```
we do not need to check for log_match, it will fail if no match found.

anamikau commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+
+    def set_primetime(self, ptime_start, ptime_end):
+        # Delete all entries in the holidays file
+        self.scheduler.holidays_delete_entry('a')
```
This method is defined in between which looks awkward. can you move at the beginning of the class.

anamikau commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+        of less than or equal to 4:00:00.
+        """
+        a = {'resources_available.ncpus': 1}
+        self.server.manager(MGR_CMD_SET, NODE, a, id=self.mom.shortname)
```
This is common to most of the test cases. please move this to setUp.

anamikau commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+        attr = {'Resource_List.walltime': (GE, '00:01:00')}
+        self.server.expect(JOB, attr, id=j2id)
+
+    def submit_resv(self, resv_start, ncpus, resv_dur):
```
same here. please move this at the beginning of the class.

anamikau commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, qname)
+        
+        a = {'enabled': 'True', 'started': 'True', 'priority': '150'}
+        self.server.manager(MGR_CMD_SET, QUEUE, a, qname)
```
can above be combined in line 366-367

anamikau commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+        self.server.expect(JOB, {'job_state': 'R'}, id=j2id)
+        self.server.expect(JOB, {'job_state': 'S'}, id=jid)
+
+        time.sleep(5)
```
do you know why are we sleeping here?

bhroam replied:
I added this sleep to allow the preempted job to run for a little while.  After reviewing it, the initial 65s sleep should handle what the test case requires.  This test fails due to PP-1049.  I am now skipping it.

anamikau commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+        self.assertTrue(len(m) > 0)
+
+    def submit_jq(self, ncpus):
+        a = {'Resource_List.select': '1:ncpus=%d' % ncpus,
```
move this to top of the class

anamikau commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+            self.assertTrue('Resource limits can not be set for the resource'
+                            in e.msg[0])
+
+    def test_t6(self):
```
This test looks good. I would suggest adding another test where setting min_walltime greater than max_walltime and making sure max_walltime is not unset.

bhroam replied:
I extended the test to set min_walltime > the existing max_walltime and setting them both at the same time.

anamikau commented:
@bhroam , please add the latest test execution results for soft walltime and STF test plan to the JRIA ticket.

bhroam commented on test/tests/functional/pbs_soft_walltime.py:
```diff
@@ -275,20 +267,20 @@ def test_soft_walltime_extend(self):
         jid = self.server.submit(J)
 
-        self.server.alterjob(jid, {'Resource_List.soft_walltime': 5})
+        self.server.alterjob(jid, {'Resource_List.soft_walltime': 6})
```
I think this will fix the issue you were seeing with this test.  The job was missing the 15s extension due to the expect calls.  Each expect call added 2s.  6s+6s+4s = 16s.  This caused us to miss the 15s extension and jump straight to 20s.  With 6s, we should have more leeway for the test to run.  If you still see this test fail, we can extend it further.

bhroam commented on test/tests/functional/pbs_stf.py:
```diff
@@ -0,0 +1,761 @@
+
+        # Make walltime given by qstat to agree with format in sched_logs.
+        # A leading '0' is removed in the hour string.
+        hh = wt.split(':')[0]
```
It won't matter.  PBS's walltime type resources use an encode function which turns the internal number into HH:MM:SS.  It isn't qstat that is doing the conversion, the server is doing it.

arungrover replied:
got it

arungrover commented:
Change looks good @bhroam

minghui-liu commented on src/server/resc_attr.c:
```diff
@@ -907,3 +908,203 @@ int apply_select_inchunk_rules(resource *presc, attribute *pattr, void *pobj, in
+						return PBSE_MIN_GT_MAXWT;
+				}
+			} else
+				return PBSE_MAX_NO_MINWT;
```
Please wrap the else case with brackets.

minghui-liu commented on src/server/req_modify.c:
```diff
@@ -483,13 +452,15 @@ modify_job_attr(job *pjob, svrattrl *plist, int perm, int *bad)
 	int	   modified = 0;
-	attribute  newattr[(int)JOB_ATR_LAST];
+	attribute *newattr;
+	attribute *newattr2;
```
Please use a more meaningful variable name.
